// react-axios/lib/components/Request.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _axios2 = require('axios');

var _axios3 = _interopRequireDefault(_axios2);

var _utils = require('../utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Request = function (_React$Component) {
  _inherits(Request, _React$Component);

  function Request(props) {
    _classCallCheck(this, Request);

    var _this2 = _possibleConstructorReturn(this, _React$Component.call(this, props));

    _this2.state = {
      isLoading: false,
      response: null,
      error: null
    };
    // create debounce function
    _this2.setupDebounce(props);
    return _this2;
  }

  Request.prototype.componentWillReceiveProps = function componentWillReceiveProps(newProps) {
    if (this.props.debounce !== newProps.debounce) {
      this.setupDebounce(newProps);
    }

    // quick and dirty prop compare
    var oldPropStr = JSON.stringify(this.props);
    var newPropStr = JSON.stringify(newProps);
    if (oldPropStr != newPropStr && newProps.isReady) {
      this.debounceMakeRequest(this.getConfig(newProps));
    }
  };

  Request.prototype.componentDidMount = function componentDidMount() {
    if (this.props.isReady) {
      this.debounceMakeRequest(this.getConfig(this.props));
    }
  };

  Request.prototype.setupDebounce = function setupDebounce(props) {
    this.debounceMakeRequest = (0, _utils.debounce)(this.makeRequest, props.debounce);
  };

  Request.prototype.getConfig = function getConfig(props) {
    return Object.assign({ url: props.url, method: props.method, data: props.data }, props.config);
  };

  Request.prototype.makeRequest = function makeRequest(config) {
    var _axios = this.props.instance || this.context.axios || _axios3.default;
    // setup cancel tokens
    if (this.source) {
      this.source.cancel('Canceling previous request.');
    }
    this.source = _axios3.default.CancelToken.source();

    // set the isLoading flag
    this.setState({ isLoading: true, error: null });
    if (typeof this.props.onLoading === 'function') {
      this.props.onLoading();
    }

    // time to make the axios request
    var _this = this;
    _axios.request(Object.assign({ cancelToken: this.source.token }, config)).then(function (res) {
      _this.setState({ isLoading: false, response: res });
      if (typeof _this.props.onSuccess === 'function') {
        _this.props.onSuccess(res);
      }
    }, function (err) {
      if (!_axios.isCancel(err)) {
        _this.setState({ isLoading: false, response: null, error: err });
        if (typeof _this.props.onError === 'function') {
          _this.props.onError(err);
        }
      }
    });
  };

  Request.prototype.render = function render() {
    if (typeof this.props.children === 'function') {
      return this.props.children(this.state.error, this.state.response, this.state.isLoading);
    }
    return null;
  };

  return Request;
}(_react2.default.Component);

Request.contextTypes = {
  axios: _react2.default.PropTypes.func
};

Request.defaultProps = {
  url: '',
  method: 'get',
  data: {},
  config: {},
  debounce: 200,
  isReady: true
};

Request.propTypes = {
  instance: _react2.default.PropTypes.func,
  url: _react2.default.PropTypes.string.isRequired,
  method: _react2.default.PropTypes.string.isRequired,
  data: _react2.default.PropTypes.object,
  config: _react2.default.PropTypes.object,
  isReady: _react2.default.PropTypes.bool,
  debounce: _react2.default.PropTypes.number,
  onSuccess: _react2.default.PropTypes.func,
  onLoading: _react2.default.PropTypes.func,
  onError: _react2.default.PropTypes.func,
  children: _react2.default.PropTypes.func
};

exports.default = Request;
module.exports = exports['default'];
